magento/magento2
# Free Shipping Patch fixes

Cart price rules for free shipping are broken on all shipping methods including Free Shipping.  Currently cart rules do not work at all properly on UPS, FedEx or USPS.  On the Free Shipping Method either the entire shipment is free or none of it.  You cannot allow only free shipping on selected items (contrary to what is shown as possible in the cart rules generation page).


### Description
This patch takes the pull request #11502 of @JosephMaxwell and fixes it such that free shipping works on select items only.  It also changes the PR to properly charge for non-free items and not change the rate charges as free items are added and removed from the order.

### Fixed Issues (if relevant)

1. magento/magento2#12573 Free Shipping Shopping Cart Rule doesn't work with UPS
2. magento/magento2#14206 Free Shipping For Matching Items Only
3. magento/magento2#10704 Free shipping code is not working for UPS for magento 2
### Manual testing scenarios
1. Generate a Cart price rule for free shipping on selected items (or subtotal of selected items).
2. Add free shipping and non free shipping items to cart.
3. Go to checkout and see proper shipping charges (or free charges) depending on items added to cart.

### Contribution checklist
